src/parquet/file/metadata.h

```diff
@@ -0,0 +1,211 @@
+      const uint8_t* metadata);
+
+  explicit RowGroupMetaDataReader(const uint8_t* metadata);
+  ~RowGroupMetaDataReader();
+  // row-group metadata
+  int num_columns() const;
+  int64_t num_rows() const;
+  int64_t total_byte_size() const;
+  // column chunk
+  int64_t column_file_offset(int i) const;
+  const std::string& column_file_path(int i) const;
+  // column metadata
+  Type::type column_type(int i) const;
+  int64_t num_values(int i) const;
+  const std::shared_ptr<schema::ColumnPath> column_path_in_schema(int i) const;
+  bool IsColumnStatsSet(int i) const;
+  ColumnStatistics GetColumnStats(int i) const;
+  Compression::type GetColumnCompression(int i) const;
+  std::vector<Encoding::type> GetColumnEncodings(int i) const;
+  int64_t GetColumnCompressedSize(int i) const;
+  int64_t GetColumnUnCompressedSize(int i) const;
+
+ private:
+  // PIMPL Idiom
+  class InternalRowGroupMetaDataReader;
+  std::unique_ptr<InternalRowGroupMetaDataReader> impl;
+};
+
+class PARQUET_EXPORT FileMetaDataReader {
+ public:
+  // API convenience to get a MetaData reader
+  static std::unique_ptr<FileMetaDataReader> GetFileMetaDataReader(
+      const uint8_t* metadata);
+
+  explicit FileMetaDataReader(const uint8_t* metadata);
+  ~FileMetaDataReader();
+  // file metadata
+  int64_t num_rows() const;
+  int num_row_groups() const;
+  int32_t version() const;
+  const std::string& created_by() const;
+  int num_schema_elements() const;
+  std::unique_ptr<SchemaMetaDataReader> GetSchemaMetaDataReader(int i) const;
+  std::unique_ptr<RowGroupMetaDataReader> GetRowGroupMetaDataReader(int i) const;
+
+ private:
+  // PIMPL Idiom
+  class InternalFileMetaDataReader;
+  std::unique_ptr<InternalFileMetaDataReader> impl;
+};
+
+class PARQUET_EXPORT SchemaMetaDataWriter {
+ public:
+  // API convenience to get a MetaData reader
+  static std::unique_ptr<SchemaMetaDataWriter> GetSchemaMetaDataWriter();
+
+  SchemaMetaDataWriter();
+  ~SchemaMetaDataWriter();
+  const uint8_t* get_buffer();
+  size_t get_buffer_size();
+
+  // SchemaElement metadata
+  void PhysicalType(Type::type ptype);
+  void type_length(int32_t length);
+  void num_children(int32_t children);
+  void scale(int32_t scale);
+  void precision(int32_t precision);
+  void field_id(int32_t id);
+  void name(const std::string& name);
+  void ConvertedType(LogicalType::type logical_type);
+  void RepetitionType(Repetition::type repetition_type);
+
+ private:
+  // PIMPL Idiom
+  class InternalSchemaMetaDataWriter;
+  std::unique_ptr<InternalSchemaMetaDataWriter> impl;
+};
+
+class PARQUET_EXPORT RowGroupMetaDataWriter {
+ public:
+  // API convenience to get a MetaData reader
+  static std::unique_ptr<RowGroupMetaDataWriter> GetRowGroupMetaDataWriter(
+      const WriterProperties& props);
+
+  explicit RowGroupMetaDataWriter(const WriterProperties& props);
+  ~RowGroupMetaDataWriter();
+  const uint8_t* get_buffer();
+  size_t get_buffer_size();
```

Comment:
Per comments elsewhere, if we make this into a more builder-like API with a `Finish` method, then the `RowGroupMetaDataBuilder` should have a reference to the `FileMetaDataBuilder` parent so that these methods are not necessary